python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -0,0 +1,861 @@
+        node.
+
+        Raises
+        ------
+        NotImplementedError if we couldn't evaluate things. Ideally
+        this should not occur, since the translation phase should pick
+        up things that we cannot handle.
+        """
+        raise NotImplementedError
+
+
+@dataclass(slots=True)
+class PythonScan(IR):
+    """Representation of input from a python function."""
+
+    options: Any
+    """Arbitrary options."""
+    predicate: Expr | None
+    """Filter to apply to the constructed dataframe before returning it."""
+
+
+@dataclass(slots=True)
+class Scan(IR):
+    """Input from files."""
+
+    typ: Any
+    """What type of file are we reading? Parquet, CSV, etc..."""
+    paths: list[str]
+    """List of paths to read from."""
+    file_options: Any
+    """Options for reading the file.
+
+    Attributes are:
+    - ``with_columns: list[str]`` of projected columns to return.
+    - ``n_rows: int``: Number of rows to read.
+    - ``row_index: tuple[name, offset] | None``: Add an integer index
+        column with given name.
+    """
+    predicate: Expr | None
+    """Mask to apply to the read dataframe."""
+
+    def __post_init__(self):
+        """Validate preconditions."""
+        if self.file_options.n_rows is not None:
+            raise NotImplementedError("row limit in scan")
+        if self.typ not in ("csv", "parquet"):
+            raise NotImplementedError(f"Unhandled scan type: {self.typ}")
+
+    def evaluate(self, *, cache: dict[int, DataFrame]) -> DataFrame:
+        """Evaluate and return a dataframe."""
+        options = self.file_options
+        n_rows = options.n_rows
+        with_columns = options.with_columns
+        row_index = options.row_index
+        assert n_rows is None
+        if self.typ == "csv":
+            df = DataFrame.from_cudf(
+                cudf.concat(
+                    [cudf.read_csv(p, usecols=with_columns) for p in self.paths]
+                )
+            )
+        elif self.typ == "parquet":
+            df = DataFrame.from_cudf(
+                cudf.read_parquet(self.paths, columns=with_columns)
+            )
+        else:
+            assert_never(self.typ)
+        if row_index is not None:
+            name, offset = row_index
+            dtype = self.schema[name]
+            step = plc.interop.from_arrow(pa.scalar(1), data_type=dtype)
+            init = plc.interop.from_arrow(pa.scalar(offset), data_type=dtype)
+            index = Column(
+                plc.filling.sequence(df.num_rows(), init, step), name
+            ).set_sorted(
+                is_sorted=plc.types.Sorted.YES,
+                order=plc.types.Order.ASCENDING,
+                null_order=plc.types.null_order.AFTER,
+            )
+            df = df.with_columns([index])
+        # TODO: should be true, but not the case until we get
+        # cudf-classic out of the loop for IO since it converts date32
+        # to datetime.
+        # assert all(
```

Comment:
My approach was going to be, once pylibcudf is spun out as a package, try installing in the environment no cudf and see what breaks. Until that is the case `cudf` will always be in globals because pylibcudf lives underneath.